### ISSUE TITLE
Change documentation structure

### ISSUE THREAD

**Issue by lock9** [neo-project/docs#685]
Hello, 
I would like to start a discussion about changing the current documentation structure. 
Here is how it is today:
https://www.dropbox.com/s/xjrvx037yjtz7iz/current-docs.png?dl=0

I think that it is not very well "organised", I don't know exactly what is wrong, but I can say that, from my experience, that people find the documentation too hard to understand or the information does not exist. 

I would like to propose the following structure:
https://www.dropbox.com/s/9zjmgwribxrn1ee/proposed-docs.png?dl=0

In text:
```
About NEO
	NEO
		NEO Foundation
		NGD - NEO Global Development 
		NGC - NEO Global Capital
	NEO Technology
		Consensus Mechanism
			Consensus Whitepaper
			Consensus Diagram
		Whitepaper
		Network Components
			Consensus Nodes
			Relay Nodes
			Light Wallets
		NEO 3.0
	NEO Virtual Machine
	NEO .NET Dev Pack
	NEO Enhancement Proposal (NEP)
	NEO Community
		NEO Research
		New Econo Lab (NEL)
		City of Zion (CoZ)
		(Add other)

Smart Contract Development
	Setting up your development environment
	Smart Contract Basics
		Supported types
			Boolean, byte, integer, string, array
		Contract Triggers
		System Fees
		Deploying your smart contract
		Invoking a smart contract
		Retrieving notifications
	Samples and Tutorials
		C#
			Hello World
			Lock Contract
			Domain (Domain Name System)
			Contract Authentication Tutorial
			Deploying a Lock Contract
		Python
	Custom token (NEP-5)
	API Reference
	Unit Testing

Blockchain Development
	Using the SDK
		C#
		Python
		GO
	Using RPC (HTTP)
	Hosting a private chain
	Configuration Files
	Wallet file
	NEO Plugins
	Advanced Topics
		Network Protocol
		Transaction Serialization
	Document for Exchange Developers
		Version 2.9.2
		Version 2.9.1
		Version 2.9.0
		Version 2.8.0
		Version 2.7.6
	Blockchain Explorers
	Network Monitoring
		CoZ Monitor
		Happy Nodes
		Notification Server

```

I would like to invite the community to give some feedback about the current / proposed structure so we can improve the documentation, there are many projects built by the community that I don't know (I'm not even sure if there is a smart contract compiler for neo-go). I also didn't add NEOJ.

Thanks

**Comment by shargon** [comment]
Documentation is so needed, we should improve this, so... 👏 good work!

**Comment by ixje** [comment]
Nice idea! 

As for feedback;
I'd propose to move "NEO .NET Dev Pack" away from "About NEO"  and towards "Smart Contract Dev" or "Blockchain Dev", whichever fits best based on what it does. 
- One reason is that it's only applicable in development
- The second (imo more important) reason is that we need to stop positioning NEO as a C# project, but as a language agnostic project. By presenting "C#/.NET something" in an about section you're expressing a clear preference for some language. 

Even though the project might personally develop actively in C#, we should show that other languages are also kept in mind. Meaning; other languages don't just have to find a way to port code from C# to "language x", but it's being taken into account when designing features. I know this is currently not really the case (simply shown by the lack of a specification), but let's at least try to work towards it.

**Comment by Celia18305** [comment]
Thanks @lock9 for your nice proposal. That's really helpful. Given that it is a significant change, we plan to take a couple of weeks to discuss and solicit comments on the proposed strucuture in both NGD and community. So let's keep this issue open and update opinions here.

**Comment by lock9** [comment]
Hello,

@ixje  I would like to present the programming language examples following something like the Unity documentation:
https://unity3d.com/learn/tutorials/topics/scripting/c-vs-js-syntax (scroll down a little)
I proposed the devpack in "about neo" because it is something that "they advertise a lot". I consider using "regular" programming languages a good advantage, but that indeed does not refer only to .NET.

I think maybe I should divide the nodes into C# and Python. The code is quite different and they do not behave in the same way. Does neo-python supports plugins? I'm not sure.

Regarding the neo-vm:
I think it should appear in both places. It should be present in "About NEO" mostly  because of the InteropInterface. I imagined that in 'about neo'  we could give a brief description on how the VM operates, and leave the details on the advanced topics. What do you think? 
Almost everyone that I explained NEO wants to see the ApplicationEngine / ExecutionEngine class.

@Celia18305  thank you, this is just a suggestion, you can count on me to migrate this docs when you guys make a decision. 
I think the most important thing to discuss is the structure, so we can add / remove content with ease.

**Comment by ixje** [comment]
> @ixje I would like to present the programming language examples following something like the Unity documentation:

The tabbed examples approach looks really good. I like it!

> I proposed the devpack in "about neo" because it is something that "they advertise a lot". I consider using "regular" programming languages a good advantage, but that indeed does not refer only to .NET.

Agree, common programming language support can be an advantage. They also promote this multi-language support for writing smart contracts. It's just some places promote multi, others only focus on C#. I think we should try to smoothing it out to language agnostic as the end goal is.

> I think maybe I should divide the nodes into C# and Python. The code is quite different and they do not behave in the same way. Does neo-python supports plugins? I'm not sure.

w.r.t nodes; don't forget neo-go. Afaik that's also running fine (need to double check with @anthdm)
w.r.t. `neo-python` plugins. it recently started adding support and will be expanding it to other parts of the code. 

> Regarding the neo-vm:
> I think it should appear in both places. It should be present in "About NEO" mostly because of the InteropInterface. I imagined that in 'about neo' we could give a brief description on how the VM operates, and leave the details on the advanced topics. What do you think?
> Almost everyone that I explained NEO wants to see the ApplicationEngine / ExecutionEngine class.

You seem to already have a proven use-case/rationale for it, so sounds good to me 👍

**Comment by lock9** [comment]
Another thing we need to discuss, is wether to put or not dApps in this documentation.
I know we already have some kind of discussion about this in the past, I don't know what was the final result of this discussiong, but I'm in favor of adding  them.
I suggest we add "dApps" (for regular dApps - 'anyone can join' ) and "Outstanding dApps", for more noticiable projects. 
I know this can be a sensitive subject, but I don't think that hiding NEO projects is good for anyone.

**Comment by lock9** [comment]
Hello,
I updated my suggestion:
```
NEO DOCS
	About NEO
		NEO
			NEO Foundation
			NGD - NEO Global Development 
			NGC - NEO Global Capital
		NEO Technology
			Consensus Mechanism
				Consensus Whitepaper
				Consensus Diagram
			Whitepaper
			Network Components
				Consensus Nodes
				Relay Nodes
				Light Wallets
			NEO 3.0
		NEO Virtual Machine
		NEO Enhancement Proposal (NEP)
		NEO Community
			NEO Research
			New Econo Lab (NEL)
			City of Zion (CoZ)
			(Add other)
	Blockchain Development
		Using the SDK
			C#
			Python
			GO
		Using RPC (HTTP)
		Asset Transfer
			NEO
			GAS
			NEP-5
		Fees
			Network fees
			System fees
		Hosting a private chain
		Configuration Files
		Wallet file
			Wallet representation
		NEO Plugins
		Advanced Topics
			Network Protocol
			Transaction Serialization
				Contract Transaction
				Invocation Transaction
				Smart Contract deployment
			Oracles
			Using the wallet through RPC
		Document for Exchange Developers
			Version 2.9.2
			Version 2.9.1
			Version 2.9.0
			Version 2.8.0
			Version 2.7.6
		Blockchain Explorers
		Network Monitoring
			CoZ Monitor
			Happy Nodes
			Notification Server
	Smart Contract Development
		Setting up your development environment
			Local network
				neo-local
			Using the debugger
				C#
				Python
		Smart Contract Basics
			Supported types
				Boolean, byte, integer, string, array, map
			Authentication
				Check Witness
				Verify Signature
				Multisignature Contracts
			Storage
			Contract Triggers
			Hash functions
			Random numbers
			Serialization
			Fees
				System Fees
				Network Fees
			Object Serialization
			Deploying your smart contract
			Invoking a smart contract
			Retrieving notifications
		Samples and Tutorials
			Hello World
				C#
				Python
			Lock Contract
			Domain (Domain Name System)
			Contract Authentication Tutorial
			Deploying a Lock Contract
		Custom token (NEP-5)
		API Reference
		Unit Testing
		NEO .NET Dev Pack
	Downloads
		Executables
			NEO-cli
			NEO-gui
			Neon
		Configuration files
		Blockchain files
```

I added a Downloads section. I think we also are going to need some sort of "guide/walkthrough" section. What do you think? We have examples, but I think we need more steps before getting into the examples (you need everything set up).

**Comment by ixje** [comment]
```
Authentication
				Check Witness
				Verify Signature
				Multisignature Contracts
Contract Triggers
```
Nice sections! These are definitely items that raise confusion and were information is not straight forward to find (it exists, but all over the place).

w.r.t walkthrough. I think it's common to have direct API documentation (for those that have already setup an environment) as well as some walkthrough that leads you from installing the latest version/SDK to building and running something. In our case I think that should be building and deploying a smart contract

**Comment by lock9** [comment]
Hello,

1 - I imagined that Asset transfer would include NEO, GAS, CGAS and NEP-5 transfers. I think it is important to cover multiple transfer in one transaction, there isn't much documentation about it.
2 - The RPC is an important topic because it can be harmful to the users that use it without the correct understanding. For instance, we have an RPC call "dumpprivkey", if the user turn on RPC on and open a wallet, he/she can loose his/her funds. I think that the usage of RPC is not totally clear, there are 2 kinds of usage: create a "public rpc endpoint"  and using "rpc in a private environment". 
Also, the "invoke", "sendRawTransaction" etc is not covered by use cases, it is very, very common to see people trying to do "write" operations using invoke/invokeScript.
3 - Like mentioned before, RPC in closed environments is different than having your own RPC server to communicate with. If the user is not clear about this 2 use cases, there is a very high chance he  is doing it wrong. 

2 - The downloads are used to download binaries, like neo-cli, neo-gui, neo-compiler, wallets etc. I think this will help users to understand better NEO components. I think it's probably going to be the starting point of any developer. Users are probably going to download the tool before checking the docs itself. I added in different section to incentivise the user to click and explore NEO related projects.

3 - Neo local is applicable to all programming languages. It runs a blockchain and you only need the .avm to deploy it. For instance, I develop using Visual Studio, but I deploy it to neo-local using neo-python from neo-local.

These are just suggestions, feel free to change it :)

**Comment by jean-acsas** [comment]
Nice suggestions @lock9. Maybe naive question: why is Java not included? 

The official getting started docs for as seen here (checked the German version too) are deprecated: [](http://docs.neo.org/en-us/sc/quickstart/getting-started-java.html). 

They point to the old neo-java-sdk repo (f.e. to [this](https://github.com/CityOfZion/neo-java-sdk/blob/master/target/org.neo.smartcontract.framework.jar) link), whereas imho they should incorporate the work here: 
[](https://github.com/neo-project/neo-devpack-java),  [](https://github.com/cwitchger/examples-java ).

New Java devs might be confused when hitting [](http://docs.neo.org/en-us/sc/quickstart/getting-started-java.html).

**Comment by lock9** [comment]
@jean-acsas I didn't include it because I have no idea of it's current situation and examples are rare, but I think the final version should include it

**Comment by lock9** [comment]
@jean-acsas  I just found out there is no SDK for java, only there is an SDK and the "dev pack", but there is no SDK to generate addresses, wallet, get block height, etc.

**Comment by jean-acsas** [comment]
@lock9 hmm, thanks for looking into the issue. that sound kind of weak tbh. would you be interested in bootstrapping such a thing? I currently quite booked, but could contribute.

**Comment by Celia18305** [comment]
We finally decided to hold over the doc structure change as well as the final decision till next year February as NEO DevCon and Chinese Spring Festival is forthcoming, maintaining a stable doc set and familiar structure for developers is our top priority. At this stage we will still focus on existing documents improvements and meanwhile keep on collecting comments from developers. After DevCon we can get down to the doc reorganization based on all feedback we've got.
Thanks a lot for all your concerns and comments.
